A80-pid.md

```diff
@@ -0,0 +1,312 @@
+
+The `proportionalGain` and `derivativeGain` parameters are taken from the LB config. `proportionalGain` should be scaled by the `WeightUpdatePeriod` value. This is necessary because `controlErrorDerivative` is inversely proportional to the sampling interval, which in turn is close to the `WeightUpdatePeriod` as we will be updating the PID state once per `WeightUpdatePeriod`. If `WeightUpdatePeriod` is too small, `controlErrorDerivative` becomes too large and dominates the resulting control error.
+
+### Extending WRR balancer
+
+The `pid` balancer reuses 90% of the wrr code. The proposal is to refactor the `wrr` codebase and introduce several hooks that allow other balancers, like `pid`, to reuse the code efficiently without the need for duplication. Initially, these hooks will remain internal to avoid prematurely establishing a new public API. This approach is mostly language-specific, but the general plan is as follows:
+
+* Add a `callbacks` object to the `wrr` balancer: This object will contain a series of callback functions that wrr will invoke at various stages of its lifecycle.
+* Introduce a `callbackData` object: This will be utilized by the callbacks to store any data that is reused across different callback functions. The `wrr` balancer will pass this object to all callbacks and treat it as an opaque blob of data.
+
+The `callbacks` object, which is to be provided by the balancer builder, will implement the following interface (expressed in pseudo-code):
+
+```
+wrrCallbacks interface {
+  onSubchannelAdded(subchannelID int, data callbackData)
+  onSubchannelRemoved(subchannelID int, data callbackData)
+
+  // onLoadReport is called when a new load report is received for a given subchannel. 
+  // This function returns the new weight for a subchannel. If the returned value is -1,
+  // the subchannel should keep using the old value.
+  // onLoadReport won't be called during the blackout period.
+  onLoadReport(subchannelId int, data callbackData, conf lbConfig, report loadReport) float
+
+  // onEDFSchedulerUpdate is called after the wrr balancer recreates the EDF scheduler.
+  onEDFSchedulerUpdate(data callbackData)
+}
+
+// Implementation for PID balancer
+func onSubchannelAdded(subchannelID int, data callbackData) {
+  // Do nothing
+}
+
+func onSubchannelRemoved(subchannelID int, data callbackData) {
+  // Remove subchannelID from two maps that store the value of last utilization
+  // and last applied weight per subchannel
+  delete(data.utilizationPerSubchannel, subchannelID)
+  delete(data.lastAppliedWeightPerSubchannel, subchannelID)
+}
+
+
+func onLoadReport(subchannelId int, data callbackData, conf lbConfig, load loadReport, lastApplied time) float {
```

Comment:
Have we moved into implementation here? I thought we were describing the API interfaces for `wrr`.

Reply:
splitted this code block into 2 and added a note to make it more apparent that we moved to implementation.